### Experiment-3 Simple Imputer with Strategy Mean

In [38]:
from sklearnex import patch_sklearn
patch_sklearn()
from sklearn.impute import SimpleImputer

simple_imputer = SimpleImputer(strategy='mean')

X_simple_impute = simple_imputer.fit_transform(X)

Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)


In [39]:
from sklearnex import patch_sklearn
patch_sklearn()
from sklearn.preprocessing import RobustScaler

RobustScaler = RobustScaler()

X_scale = RobustScaler.fit_transform(X_simple_impute)

Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)


In [40]:
smt = SMOTETomek(random_state=42,sampling_strategy='minority', n_jobs=-1 )
X_resample, y_resample = smt.fit_resample(X_scale, y)

In [41]:
X_train,X_test,y_train,y_test = train_test_split(X_resample,y_resample,test_size=0.2,random_state=42)

In [42]:
model_logreg = LogisticRegression()
model_logreg.fit(X_train,y_train)
y_pred = model_logreg.predict(X_test)
accuracy = accuracy_score(y_test,y_pred)
confusion_mat = confusion_matrix(y_test, y_pred)
classfication_report = classification_report(y_test,y_pred)

print("The accuracy score: ",accuracy*100)
print("\n Confusion matrix:")
print(confusion_mat)
print("\n Classification Report:")
print(classfication_report)
print("\n Total Cost due to misclassification: ",total_cost(y_test,y_pred))

The accuracy score:  58.83820384889522

 Confusion matrix:
[[1926 5172]
 [ 603 6329]]

 Classification Report:
              precision    recall  f1-score   support

           0       0.76      0.27      0.40      7098
           1       0.55      0.91      0.69      6932

    accuracy                           0.59     14030
   macro avg       0.66      0.59      0.54     14030
weighted avg       0.66      0.59      0.54     14030


 Total Cost due to misclassification:  353220


In [43]:
model_knn = KNeighborsClassifier()
model_knn.fit(X_train,y_train)
y_pred = model_knn.predict(X_test)
accuracy = accuracy_score(y_test,y_pred)
confusion_mat = confusion_matrix(y_test, y_pred)
classfication_report = classification_report(y_test,y_pred)

print("The accuracy score: ",accuracy*100)
print("\n Confusion matrix:")
print(confusion_mat)
print("\n Classification Report:")
print(classfication_report)
print("\n Total Cost due to misclassification: ",total_cost(y_test,y_pred))

The accuracy score:  97.42694226657164

 Confusion matrix:
[[6821  277]
 [  84 6848]]

 Classification Report:
              precision    recall  f1-score   support

           0       0.99      0.96      0.97      7098
           1       0.96      0.99      0.97      6932

    accuracy                           0.97     14030
   macro avg       0.97      0.97      0.97     14030
weighted avg       0.97      0.97      0.97     14030


 Total Cost due to misclassification:  44770


In [44]:
model_gbc = GradientBoostingClassifier()
model_gbc.fit(X_train,y_train)
y_pred = model_gbc.predict(X_test)
accuracy = accuracy_score(y_test,y_pred)
confusion_mat = confusion_matrix(y_test, y_pred)
classfication_report = classification_report(y_test,y_pred)

print("The accuracy score: ",accuracy*100)
print("\n Confusion matrix:")
print(confusion_mat)
print("\n Classification Report:")
print(classfication_report)
print("\n Total Cost due to misclassification: ",total_cost(y_test,y_pred))

The accuracy score:  98.64575908766928

 Confusion matrix:
[[6983  115]
 [  75 6857]]

 Classification Report:
              precision    recall  f1-score   support

           0       0.99      0.98      0.99      7098
           1       0.98      0.99      0.99      6932

    accuracy                           0.99     14030
   macro avg       0.99      0.99      0.99     14030
weighted avg       0.99      0.99      0.99     14030


 Total Cost due to misclassification:  38650


In [45]:
model_xgboost = XGBClassifier()
model_xgboost.fit(X_train,y_train)
y_pred = model_xgboost.predict(X_test)
accuracy = accuracy_score(y_test,y_pred)
confusion_mat = confusion_matrix(y_test, y_pred)
classfication_report = classification_report(y_test,y_pred)

print("The accuracy score: ",accuracy*100)
print("\n Confusion matrix:")
print(confusion_mat)
print("\n Classification Report:")
print(classfication_report)
print("\n Total Cost due to misclassification: ",total_cost(y_test,y_pred))

The accuracy score:  99.70064148253742

 Confusion matrix:
[[7060   38]
 [   4 6928]]

 Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00      7098
           1       0.99      1.00      1.00      6932

    accuracy                           1.00     14030
   macro avg       1.00      1.00      1.00     14030
weighted avg       1.00      1.00      1.00     14030


 Total Cost due to misclassification:  2380


In [46]:
model_randomforest = RandomForestClassifier()
model_randomforest.fit(X_train,y_train)
y_pred = model_randomforest.predict(X_test)

accuracy = accuracy_score(y_test,y_pred)
confusion_mat = confusion_matrix(y_test, y_pred)
classfication_report = classification_report(y_test,y_pred)

print("The accuracy score: ",accuracy*100)
print("\n Confusion matrix")
print(confusion_mat)
print("\n Classification Report")
print(classfication_report)
print("\n Total Cost due to misclassification: ",total_cost(y_test,y_pred))

The accuracy score:  99.40841054882395

 Confusion matrix
[[7027   71]
 [  12 6920]]

 Classification Report
              precision    recall  f1-score   support

           0       1.00      0.99      0.99      7098
           1       0.99      1.00      0.99      6932

    accuracy                           0.99     14030
   macro avg       0.99      0.99      0.99     14030
weighted avg       0.99      0.99      0.99     14030


 Total Cost due to misclassification:  6710


### 

### The Best Model with an accuracy of 99.70% and total cost of 2380 due to misclassification

### 

### Saving the Best Model

In [48]:
import joblib

In [50]:
joblib.dump(model_xgboost,"best_model.joblib")

['best_model.joblib']